src/query/ts/m3db/consolidators/consolidator_single.go

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package consolidators
+
+import (
+	"math"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/ts"
+)
+
+// SingleLookbackConsolidator is a helper for consolidating a full single
+// series. It has some differences with the step consolidator in that it
+// collects points for a single series, and is reset when the next series
+// needs to be consolidated.
+type SingleLookbackConsolidator struct {
+	lookbackDuration time.Duration
+	stepSize         time.Duration
+	earliestLookback time.Time
+	consolidated     float64
+	datapoints       []ts.Datapoint
+	fn               ConsolidationFunc
+}
+
+// NewSingleConsolidator creates a single value consolidator used for
+// series iteration with a given lookback.
+func NewSingleConsolidator(
+	lookbackDuration, stepSize time.Duration,
+	startTime time.Time,
+	fn ConsolidationFunc,
+) *SingleLookbackConsolidator {
+	datapoints := make([]ts.Datapoint, 0, initLength)
+
+	return &SingleLookbackConsolidator{
+		lookbackDuration: lookbackDuration,
+		stepSize:         stepSize,
+		earliestLookback: startTime.Add(-1 * lookbackDuration),
+		consolidated:     math.NaN(),
+		datapoints:       datapoints,
```

Comment:
Maybe just initialize this here `datapoints: make([]ts.Datapoint, 0, initLength),`